### DISCUSSION TITLE
Suggestions and Comments

### DISCUSSION THREAD

**Discussion by 0x1000000** [0x1000000/SqExpress#2] [General]
Hi developers! If you have any suggestions or comments, do not hesitate to share them here!

**Comment by greydmar** [comment]
Hello.
Can you explain (general overview) about codegen block?
Thanks in advance.
Nice, very nice work!

**Reply by 0x1000000** [reply]
Hi @greydmar ,
The code-generation can be used in the two scenarios:
1.	You have an existing SQL Database (Only MS SQL is supported right now) and you want to generate table descriptor classes.
2.	You already have table descriptor classes and you want to have some DTOs with all the mappings.
The library NuGet package goes with a console util (.Net Core 3.1) which performs the cogenerations.
The util can be found in a NuGet package folder:
On windows
```%userprofile%\.nuget\packages\sqexpress\<version>\tools\codegen\SqExpress.CodeGenUtil.dll```
On Linux
```~/.nuget/packages/sqexpress/<version>/tools/codegen/SqExpress.CodeGenUtil.dll```

To generate table descriptors, you need to specify db connection string and an output folder
[Example](https://github.com/0x1000000/SqExpress#table-descriptors-scaffolding)

To generate DTOs, you need to specify a folder with table descriptors and an output folder (for DTOs)
[Example](https://github.com/0x1000000/SqExpress#dtos-scaffolding)

The code generation is incremental (it uses Roslyn) and that means that you can add some own code (e.g. some methods and attributes) to the generated classes.
